### PR TITLE
Client  - Outputs id api querry parameters

### LINKED ISSUE #287: Client - Add querry params for get outputs indexer api
## Description

Add optional querry parameters for indexer api's.

## Requirements

Handle Querry params for /outputs, /nfts, /aliases, /foundaries end points.

## Are you planning to do it yourself in a pull request?

Yes

### PR DESCRIPTION
**Pull request by rebinmannady93**
# Description of change

Fixes #287 
- Added query params for getting basic, nft, alias and foundry outputs.
### Added Query Params :
- address
- hasDustReturnCondition
- dustReturnAddress
- sender
- tag
- pageSize
- cursor
- stateController
- governor
- issuer

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

test_outputs_id.c

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/client/api/restful/get_outputs_id.c [hunk 1]
```diff
@@ -6,6 +6,172 @@
...
+#define OUTPUTS_QUERY_GOV_KEY "governor"
+#define OUTPUTS_QUERY_ISSUER_KEY "issuer"
+
+outputs_query_list_t *outputs_query_list_new() { return NULL; }
+
+int outputs_query_list_add(outputs_query_list_t **list, outputs_query_params_e type, char const *const param) {
+  outputs_query_list_t *next = malloc(sizeof(outputs_query_list_t));
+  if (next) {
+    next->query_item = malloc(sizeof(outputs_query_params_t));
+    next->query_item->type = type;
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c]: Check for NULL pointer is missing.

File: src/client/api/restful/get_outputs_id.c [hunk 2]
```diff
@@ -6,6 +6,172 @@
...
+      case QUERY_PARAM_ISSUER:
+        query_str_len += strlen(OUTPUTS_QUERY_ISSUER_KEY);
+        query_str_len += strlen(elm->query_item->param);
+        query_str_len += 2;  // For "&" params seperator and "=" params assignment
+        break;
+      default:
+        break;
+    }
+  }
+  query_str_len--;  // Remove the "&" params seperator at the end
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c]: What if there is no query in a list?

File: src/client/api/restful/get_outputs_id.c [hunk 3]
```diff
@@ -6,6 +6,172 @@
...
+  return query_str_len;
+}
+
+static int copy_param_to_buf(char *buf, size_t offset, char *key, outputs_query_list_t *elm) {
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c]: Is `offset` argument really necessary?

File: src/client/api/restful/get_outputs_id.c [hunk 4]
```diff
@@ -6,6 +6,172 @@
...
+        offset += copy_param_to_buf(buf, offset, OUTPUTS_QUERY_GOV_KEY, elm);
+        break;
+      case QUERY_PARAM_ISSUER:
+        offset += copy_param_to_buf(buf, offset, OUTPUTS_QUERY_ISSUER_KEY, elm);
+        break;
+      default:
+        break;
+    }
+  }
+  buf[offset - 1] = 0;  // Replace the "&" at the end with '\0'
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c]: What if query list is empty?

File: src/client/api/restful/get_outputs_id.c [hunk 5]
```diff
@@ -158,109 +324,172 @@ static int get_outputs_api_call(iota_client_conf_t const *conf, char *cmd_buffer
...
+    free(query_str);
+  } else {
+    cmd_buffer = malloc(api_path_len + 1);  // api_path + '\0'
+    // copy api path
+    memcpy(cmd_buffer, INDEXER_NFT_API_PATH, api_path_len + 1);
+  }
+
+  int ret = get_outputs_api_call(conf, cmd_buffer, res);
+  free(cmd_buffer);
+  return ret;
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c]: Is this the same code as for `INDEXER_OUTPUTS_API_PATH`? Could this be somehow extracted into a new function?

**Review by matjazv — CHANGES_REQUESTED** [review]

File: src/client/api/restful/get_outputs_id.c [hunk 6]
```diff
@@ -6,6 +6,172 @@
...
+#define OUTPUTS_QUERY_ISSUER_KEY "issuer"
+
+outputs_query_list_t *outputs_query_list_new() { return NULL; }
+
+int outputs_query_list_add(outputs_query_list_t **list, outputs_query_params_e type, char const *const param) {
+  outputs_query_list_t *next = malloc(sizeof(outputs_query_list_t));
+  if (next) {
+    next->query_item = malloc(sizeof(outputs_query_params_t));
+    next->query_item->type = type;
+    next->query_item->param = malloc(strlen(param) + 1);
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.c]: the allocation might fail.

File: src/client/api/restful/get_outputs_id.c [hunk 7]
```diff
@@ -158,109 +324,172 @@ static int get_outputs_api_call(iota_client_conf_t const *conf, char *cmd_buffer
...
+      printf("[%s:%d]: OOM\n", __func__, __LINE__);
+      return -1;
+    }
+    size_t len = get_outputs_query_str(list, query_str, query_str_len + 1);
+    if (len != query_str_len + 1) {
+      printf("[%s:%d]: Query string len and copied data mismatch\n", __func__, __LINE__);
+      free(query_str);
+      return -1;
+    }
+    cmd_buffer = malloc(api_path_len + 1 + query_str_len + 1);  // api_path + '?' + query_str + '\0'
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.c:364]: `cmd_buffer` could be NULL, same to all `cmd_buffer` in this source.

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/client/api/restful/get_outputs_id.c (hunk 6 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: NULL check added.

File: src/client/api/restful/get_outputs_id.c (hunk 7 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c:364]: NULL checks added.

File: src/client/api/restful/get_outputs_id.c (hunk 1 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: NULL check added.

File: src/client/api/restful/get_outputs_id.c [hunk 8]
```diff
@@ -25,8 +25,16 @@ int outputs_query_list_add(outputs_query_list_t **list, outputs_query_params_e t
...
+      return -1;
+    }
     next->query_item->type = type;
     next->query_item->param = malloc(strlen(param) + 1);
+    if (!next->query_item->param) {
+      printf("[%s:%d]: OOM\n", __func__, __LINE__);
+      return -1;
+    }
     memcpy(next->query_item->param, param, strlen(param) + 1);
     if (next->query_item) {
```
Comment (matjazv) [inline src/client/api/restful/get_outputs_id.c:33]: This check is probably redundant because if `query_item` is NULL a function returns earlier.

**Review by matjazv — CHANGES_REQUESTED** [review]

File: src/client/api/restful/get_outputs_id.c (hunk 8 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c:33]: Yes, removed it.

File: src/client/api/restful/get_outputs_id.c (hunk 2 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: Added a check for query_str_len. If no item present in list, it return length 0.
if (query_str_len > 0) {
    query_str_len--;  // Remove the "&" params seperator at the end
  }

File: src/client/api/restful/get_outputs_id.c (hunk 4 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: Added check for query_str_len equals zero at the begining.
Also added check for buf[offset-1] == '&' at the end. If list is empty it returns 0.

File: src/client/api/restful/get_outputs_id.c (hunk 3 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: Removed offset argunemt from copy_param_to_buf() function.

File: src/client/api/restful/get_outputs_id.c (hunk 5 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: Done.

File: src/client/api/restful/get_outputs_id.h [hunk 9]
```diff
@@ -11,6 +11,46 @@
...
+  QUERY_PARAM_TAG,            ///< A tag block to search for outputs matching it
+  QUERY_PARAM_PAGE_SIZE,      ///< The maximum amount of items returned in one api call
+  QUERY_PARAM_CURSOR,         ///<  A cursor to start the query (confirmationMS+outputId.pageSize)
+  QUERY_PARAM_STATE_CTRL,     ///< To query outputs based on bech32-encoded state controller address
+  QUERY_PARAM_GOV,            ///< To query outputs based on bech32-encoded governor (governance controller) address
+  QUERY_PARAM_ISSUER          ///< To query outputs based on bech32-encoded issuer address
+} outputs_query_params_e;
+
+/**
+ * @brief A Query Param Onject
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.h]:
```suggestion
 * @brief A Query Param Object
```

File: src/client/api/restful/get_outputs_id.h [hunk 10]
```diff
@@ -38,6 +78,47 @@ typedef struct {
...
 
+/**
+ * @brief New outputs query params list
+ *
+ * @return outputs_query_list_t*
+ */
+outputs_query_list_t *outputs_query_list_new();
+
+/**
+ * @brief Add a querry parameter to the list
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.h]:
```suggestion
 * @brief Add a query parameter to the list
```

File: src/client/api/restful/get_outputs_id.h [hunk 11]
```diff
@@ -11,6 +11,46 @@
 #include "core/address.h"
 #include "core/types.h"
 
+#define INDEXER_OUTPUTS_API_PATH "/api/plugins/indexer/v1/outputs"
+#define INDEXER_ALIASES_API_PATH "/api/plugins/indexer/v1/aliases"
+#define INDEXER_NFT_API_PATH "/api/plugins/indexer/v1/nfts"
+#define INDEXER_FOUNDRY_API_PATH "/api/plugins/indexer/v1/foundries"
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.h]: these are used in `get_outputs_id.c` only, should not put in the header.

File: tests/client/api_restful/test_outputs_id.c [hunk 12]
```diff
@@ -13,6 +13,42 @@ void setUp(void) {}
...
 
+void test_query_params() {
+  char addr[] = "atoi1qpl4a3k3dep7qmw4tdq3pss6ld40jr5yhaq4fjakxgmdgk238j5hzsk2xsk";
+  char const* const cursor = "6209d527453cf2b9896146f13fbef94f66883d5e4bfe5600399e9328655fe0850fd3d05a0000.2";
+  char const* const expected_query_str =
+      "address=atoi1qpl4a3k3dep7qmw4tdq3pss6ld40jr5yhaq4fjakxgmdgk238j5hzsk2xsk&pageSize=2&cursor="
+      "6209d527453cf2b9896146f13fbef94f66883d5e4bfe5600399e9328655fe0850fd3d05a0000.2";
+  outputs_query_list_t* list = outputs_query_list_new();
+  TEST_ASSERT_NULL(list);
+  // Check len qyery string len for empty list
```
Comment (oopsmonk) [inline tests/client/api_restful/test_outputs_id.c]:
```suggestion
  // Check query string len with an empty list
```

File: src/client/api/restful/get_outputs_id.c [hunk 13]
```diff
@@ -158,109 +339,114 @@ static int get_outputs_api_call(iota_client_conf_t const *conf, char *cmd_buffer
...
+    }
+    // copy api path
+    memcpy(cmd_buffer, api_path, api_path_len);
+    // add "?" query symbol
+    cmd_buffer[api_path_len] = '?';
+    // copy query strings
+    memcpy(cmd_buffer + api_path_len + 1, query_str, query_str_len + 1);
+    free(query_str);
+  } else {
+    printf("Reached here\n");
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.c]: remove
```suggestion

```

File: tests/client/api_restful/test_outputs_id.c [hunk 14]
```diff
@@ -87,183 +123,461 @@ void test_deser_outputs_err() {
   res_outputs_free(res);
 }
 
-void test_get_output_ids_from_address() {
+void test_get_output_ids() {
   char addr[] = "atoi1qpl4a3k3dep7qmw4tdq3pss6ld40jr5yhaq4fjakxgmdgk238j5hzsk2xsk";
+  char dust_ret_addr[] = "atoi1qpszqzadsym6wpppd6z037dvlejmjuke7s24hm95s9fg9vpua7vluehe53e";
   char const* const addr_hex_invalid = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
   char const* const addr_hex_invalid_length =
       "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
```
Comment (oopsmonk) [inline tests/client/api_restful/test_outputs_id.c]: invalid length shuold use valid chracters with invalid length, right? 
like, `zpk6m4x7m2t6k5pvgs0yd2nqelfaz09ueyyv6fwn` or `atoi1zpk6m4x7m2t6k5pvgs0yd2nqelfaz09ueyyv6fwnzzzzzzzz`

File: src/client/api/restful/get_outputs_id.h [hunk 15]
```diff
@@ -11,6 +11,46 @@
...
+
+/**
+ * @brief All Query Params Type
+ *
+ */
+typedef enum {
+  QUERY_PARAM_ADDRESS = 0,    ///< The Bech32-encoded address that should be used to query outputs
+  QUERY_PARAM_DUST_RET,       ///< The presence of dust return unlock condition
```
Comment (oopsmonk) [inline src/client/api/restful/get_outputs_id.h]: more specifically, it's a boolean.
```suggestion
  QUERY_PARAM_HAS_DUST_RET,       ///< The presence of dust return unlock condition
```

File: tests/client/api_restful/test_outputs_id.c [hunk 16]
```diff
@@ -87,183 +123,461 @@ void test_deser_outputs_err() {
...
   TEST_ASSERT_NOT_NULL(res);
+  list = outputs_query_list_new();
+  TEST_ASSERT_NULL(list);
+  TEST_ASSERT(outputs_query_list_add(&list, QUERY_PARAM_ADDRESS, addr_nft) == 0);
+  TEST_ASSERT_EQUAL_INT(0, get_nft_outputs(&ctx, list, res));
+  TEST_ASSERT(res->is_error == false);
 
   //=====Test valid nft address=====
-  int ret = get_outputs_from_nft_address(&ctx, addr_nft, res);
+  outputs_query_list_free(list);
```
Comment (oopsmonk) [inline tests/client/api_restful/test_outputs_id.c]: the same test with [the above one](https://github.com/iotaledger/iota.c/pull/302/files#diff-2e2dcba4df2fd634757a11863c1230b64f647c143745b1390e44056a411cb97dR314)?

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/client/api/restful/get_outputs_id.h (hunk 9 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.h]: Done.

File: src/client/api/restful/get_outputs_id.h (hunk 11 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.h]: Done.

File: src/client/api/restful/get_outputs_id.c (hunk 13 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.c]: Done.

File: src/client/api/restful/get_outputs_id.h (hunk 15 above)
Comment (rebinmannady93) [reply src/client/api/restful/get_outputs_id.h]: Done.

File: tests/client/api_restful/test_outputs_id.c (hunk 16 above)
Comment (rebinmannady93) [reply tests/client/api_restful/test_outputs_id.c]: Removed.

**Review by matjazv — APPROVED** [review]

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**